DOCS-9268 added the policy (#295)
mulesoft/docs-api-manager: modules/ROOT/pages/rate-limiting-and-throttling-sla-based-policies.adoc

include::_attributes.adoc[]
endif::[]

The Rate Limiting policies based on a service level access (SLA) are client ID-based policies that use the client ID as a reference to impose limits on the number of requests that each application can make within a period of time. To use these policies you need to create at least one SLA tier to define request limits.
:keywords: rate limiting SLA, api gateway, gateway, policy

You can configure an SLA tier to require manual approval of requests from users who want to register an app to call your API.

> == Rate Limiting - SLA-Based Policy

Reviewer: I looked up "rate limiting sla based policy" and saw that all of the entries are from MuleSoft, and that the majority of presentations were "Rate Limiting - SLA-Based Policy."

However, this makes no logical grammatical sense to me.

"rate limiting" should be two words as a noun, but hyphenated as a compound adjective: rate-limiting policy.

"SLA-based" should be hyphenated, of course.

I would use "Rate Limiting - SLA-Based" only if it were the actual policy name, verbatim; otherwise, it makes grammatical sense for this to be "Rate-Limiting SLA-Based Policy."

Ugh.

Author: Done! Thanks!

> >s| Policy name | Rate limiting SLA

Reviewer: ```suggestion
>s| Policy name | Rate-limiting SLA
```

Author: Done! Thanks!

> >s| Summary      | Monitors access to an API by defining a maximum amount of requests processed for a period of time based on SLAs

Reviewer: NUMBER OF REQUESTS, not amount of requests.
GLOBAL
I will stop marking this from here on in this PR; please search for and change.

Reviewer: ```suggestion
>s| Summary      | Monitors access to an API by defining a maximum number of requests processed during a period of time, based on SLAs
```

Author: Done! Thanks!

>s| Category | Quality of Service
>s| First Mule version available | v4.1.0
.4+>.^s| Returned Status Codes

> |400 - Request is blocked because quota exceeded by WSDL APIs that use SOAP v1.2

Reviewer 2: This status code is the error status code for SOAP v1.2, error can be a quota exceed OR an invalid client. The later error type is not explained here.

Author: Done! Thanks!

> |401 - Request is blocked due to invalid client credentials (client ID or client secret) provided

Reviewer: Parallel syntax:
```suggestion
|401 - Request blocked due to invalid client credentials (client ID or client secret)
```

Author: Done! Thanks!

> |429 - Quota exceeded, requests are blocked until the current window finishes

Reviewer: ditto
```suggestion
|429 - Quota exceeded; requests blocked until the current window finishes
```
I still do not like the wording "a window finishes". The window does not finish; the processing taking place in the window finishes. But I will stop marking this; search for and address GLOBALLY.

Reviewer: Many of these edits repeat edits that I made in the file you sent yesterday, so please take note of the rewording and presentation suggestions for future use.

Author: Done! Thanks!

> |500 - Request is blocked because quota exceeded by WSDL APIs that use SOAP v1.1

Reviewer 2: This status code is the error status code for SOAP v1.1, error can be a quota exceed OR an invalid client. The later error type is not explained here.

Author: Done! Thanks!

|===

> The Rate Limiting service level agreement (SLA) policy enables you to control the incoming traffic to an API by limiting the number of requests that the API can receive within a given period of time. After the limit is reached, the policy rejects all requests, thereby avoiding any additional load on the backend API. 

Reviewer: If this is the actual policy name, then it should be completely init. capped: Rate Limiting SLA policy or Rate Limiting Service Level Agreement policy.

Grammatically, both "rate-limiting" and "service-level" should be hyphenated.

This would be MY solution:
```suggestion
The Rate-Limiting Service-Level Agreement (SLA) policy enables you to control incoming traffic to an API by limiting the number of requests that the API can receive within a given timespan. When the limit is reached before the time expires, the policy rejects all requests, thereby avoiding any additional load on the backend API. 
```

Reviewer: I am going to stop commenting on the presentation of the policy name.

Author: Done! Thanks!

> To apply the Rate Limiting SLA policy to an API, you must first create a contract between the API and a registered client application. The number of requests that an API can receive within a given period of time is defined in the xref:manage-client-app-taks.adoc[contracts] section in API Manager. 

Reviewer: "period of time" is verbose; "time" alone suffices, or "timespan," if it is necessary to emphasize the duration.
GLOBAL.

Author: Done! Thanks!

> Each request must be identified by a client ID and an optional client secret (depending on the policy configuration). To review how to obtain the client credentials of a registered Client application, see xref:access-client-app-id-task.adoc[Obtaining Client Credentials of a Registered Application]. 

Reviewer: GLOBALLY lowercase "client application":
```suggestion
Each request must be identified by a client ID and an optional client secret (depending on the policy configuration). To review how to obtain the client credentials of a registered client application, see xref:access-client-app-id-task.adoc[Obtaining Client Credentials of a Registered Application]. 
```

Author: Done! Thanks!

> Additionally, you can further configure the policy to run in a Mule runtime engine (Mule) cluster, or add identifiers to define groups of requests. In a clusterized policy configuration, the quota is shared between all the nodes in the cluster. For more information about how these options work, see <<examples,Examples>>.

Reviewer: "Additionally" and "further" are redundant
Replace "between" with "among":
```suggestion
You can additionally configure the policy to run in a Mule runtime engine (Mule) cluster or add identifiers to define groups of requests. In a clusterized policy configuration, the quota is shared among all nodes in the cluster. For more information about how these options work, see <<examples,Examples>>.
```

Author: Done! Thanks!

== How This Policy Works

> The Rate Limiting SLA policy keeps track of the number of requests made in the current window (the available quota), allowing the requests to reach the backend only if the available quota is greater than zero. 

Reviewer: Per PR edit yesterday, replace "keeps track of" with a less idiomatic term, such as "monitors."

Author: Done! Thanks!

> Because each client defines a separate available quota for their window, the client application must define an SLA with the API (a contract). Therefore, to verify whether the request is within the SLA limit, you must define a way to obtain the Client ID from the request, and optionally the Client Secret. 

Reviewer: ditto edits to this same text in PR #293

Reviewer: Lowercase both "client ID" and "client secret"

Author: Thanks!

> After a contract is defined between a client application and an API, API gateway automatically manages these contracts by keeping track of your API Manager configurations. Additionally, API gateway implements high availability strategies in case an unexpected downtime occurs in the Anypoint Platform management plane.

Reviewer: "define" is not the correct verb to use with "contract": a contract is an agreement, and one does not "define" an agreement but rather reach an agreement or simply agree. "made" or "created" would be better.

Reviewer: ditto "keeping track of"

Reviewer: hyphenate "high-availability strategies"

Reviewer: "in case an unexpected downtime occurs in" > "in case of unexpected downtime in"

Author: Done! Thanks!

> To understand how the Rate Limiting SLA policy works, let’s take an example of how the configuration of an SLA of 3 requests every 10 seconds for the client with ID “ID#1” allows or restricts the request, based on the quota available in that window:

Reviewer: Ditto yesterdays edits about NOT using "let's."
Please search for GLOBALLY and replace with "consider" language.

Reviewer: Also, it is incorrect English to say "Let's take an example of how...."

The proper IDIOMATIC English would be "Let's take, for example, how <this works>...." or "Taking the example of <how this works>, we can see...."

Google "take an example" and review the results.

Author: Done! Thanks!

image:rate-limiting-sla-general.png[Rate Limiting SLA-Based policy,75%,85%]

> * Requests of Client with ID “ID#1”: In the first window, because the quota is reached with the third request, all subsequent requests are rejected until the window closes. In the second window, only 2 of the 3 requests are processed. In this window, the quota remaining for the window is dropped after the window time has elapsed. 

Reviewer: This bullet-list item and the following do not follow our standard format; please correct, per style ref.

Reviewer: Spell the number of requests: "only two of the three requests"

Reviewer: Is a quota remainder "dropped" or simply left unused because the specified timespan has elapsed?

Author: Done! Thanks!

> An accepted request passes through the API to the backend. A rejected request on the other hand, displays a “429 status for HTTP, ” (or either 400 or 500 if the API is WSDL) and does not reach the backend.

Reviewer: Per the previous PR edit, DO NOT use "on the other hand." GLOBAL

[Note also that this is improperly punctuated and would be...:
[```suggestion
[An accepted request passes through the API to the backend. A rejected request, on the other hand, displays `429 status for HTTP' (or either 400 or 500 if the API is WSDL) and does not reach the backend.]
```]

Author: Done! Thanks!

> * Requests of Client with ID “ID#2”: Because the client has no contract defined for the API, every request is rejected , and therefore no request is allowed. 

Reviewer: Does not follow our standard for bullet-list style and is improperly punctuated:
* Requests of Client with ID “ID#2”
Because the client has no contract with the API, every request is rejected.

Author: Done! Thanks!

> An API may have several contracts, each with its own SLA. The Rate Limiting SLA policy keeps track of the quota (and windows) available for each client independently by creating one rate limiting algorithm per contract. With the first request to the API, algorithms are created using the _lazy creation_ strategy.

Reviewer: Same edits as yesterday:

may > might 
Replace "keeps track of"

Reviewer: The remainder of this PR is further repetition of the same text that I edited yesterday in PR #293.
Why is the same text documented in more than one place?
In future, please go through and self-edit using comments I have already made before sending to me.
The idea is to learn from edit corrections and not continue to make the same errors.

Author: thanks!

[[examples]]
== Examples

> Let’s look at how the same contract of 3 requests every 10 seconds for client with “ID#1” works when the configuration is clusterized.

Reviewer: ```suggestion
Consider how the same contract of 3 requests every 10 seconds for client ID#1 works when the configuration is clusterized.
```

Author: Done! Thanks!

> Let’s look at an example of a a window that is reset exactly at 12:00:00, with a 10 seconds length, and a 2-node Mule cluster. Both nodes start and end their windows at the same time, and the cluster allows 3 requests per window in total for a client with ID “ID#1”:

Reviewer: Look at this green text and that directly above it: they are both introductions. Delete the repetition between them.

Reviewer: Copyedit
```suggestion
Consider an example of a window that is reset exactly at 12:00:00, is 10 seconds long, and operates in a 2-node Mule cluster. Both nodes start and end their windows at the same time, and the cluster allows 3 requests per window for client ID#1:
```

Author: Done! Thanks!

image:rate-limiting-clusterizable.png[Rate Limiting SLA Clusterizalbe Configuration,75%,85%]

> For the requests of Client with ID “ID#1”: Because the policy is clusterized, the whole cluster accepts 3 requests. If the clusterizable policy is turned off, the Mule cluster can accept 6 requests per window, which amounts to 3 requests per node.

Reviewer: This text is a bullet earlier:
```suggestion
* Requests of client ID#1
Because the policy is clusterized, the whole cluster accepts three requests. If clusterization of the policy is disabled, the Mule cluster can accept six requests per window: three per node.
```

Author: Done! Thanks!

> For the requests of Client with ID “ID#2”:  Because there is no contract defined for this client in this specific API, every request is rejected.

Reviewer: ```suggestion
* Requests of client ID#2
+
Because there is no contract for this client in this specific API, every request is rejected.
```

Reviewer: The remainder of this file should take the same edits as the previous PR.
Edit ended here.

Author: Done! Thanks!
